### ISSUE TITLE
DisplayTextFor doesn't respect Display attribute

### ISSUE THREAD

**Issue by drauch** [aspnet/Mvc#7033]
Hi!

We want to render Name+Status of a list of entities in a select HTML tag like so:

    <select>
        <option value="...">Name1 (Status: Status1)</option>
        ...
    </select>

For this we create a `List<SelectListItem>` like so:

    entities.Select(e => new SelectListItem {
        Value = e.Id,
        Text = e.Name + " (Status: " + htmlHelper.DisplayTextFor(x => e.Status) + ")"
    }).ToList();

Where Status is of an enum type. The enum values have Display attributes on them, however, they are completely ignored. It looks like DisplayTextFor renders just the ToString() representation of the enum.

Is this a bug? Is this the desired behavior? What's our alternative?

We don't want to use DisplayFor() instead of DisplayTextFor() to prevent rendering HTML into the option element (which is not allowed).

**Comment by Eilon** [comment]
@dougbu / @rynowak thoughts on this?

@drauch have you tried `DisplayNameFor`?

I can't recall the exact differences between these methods...

**Comment by dougbu** [comment]
@Eilon `DisplayNameFor(...)` displays the name of a property e.g. `"Joe"` for `m => m.Fred.Barney.Joe` while `DisplayTextFor(...)` displays a string for an expression's value. (`Valuefor(...)` also displays an expression's value but it's focused on round-tripping an expression that was model bound.)

`DisplayTextFor(...)` centers on `Convert.ToString(...)`. It does have some special cases for `SimpleDisplayProperty` and `NullDisplayText`. It also displays the value (or `NullDisplayText`) of the model's first property if `Convert.ToString(...)` just display's the model type's `FullName`.

Bottom line, we've got a lot of special cases in that corner of the code and another one for `enum`s should be too painful. Would be something like the `TemplateBuilder` special case for `enum`s: https://github.com/aspnet/Mvc/blob/5fffd464cd874891a31a8cf14258aaecc099f932/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/TemplateBuilder.cs#L121-L136

**Comment by dougbu** [comment]
Clearing `investigate` label.

**Comment by drauch** [comment]
Did you intend to write "should**n't** be too painful" or is this behavior going to stay? : )

I got the facts, but I'm a little confused now on what the outcome of this is.

**Comment by dougbu** [comment]
@drauch I've updated my comment to say "should**n't**".

No outcome yet, just options and opinions.

**Comment by Eilon** [comment]
@dougbu thanks for the details!

I'm placing this in the Backlog because we have not current plans to do this, but it's listed as up-for-grabs in case someone would like to take a stab at it.

**Comment by mistakenot** [comment]
I've had an initial attempt at this in the PR above with a few comments.

**Comment by mkArtakMSFT** [comment]
@dougbu, can you please review the referenced PR? Thanks!

**Comment by dougbu** [comment]
- 4f3e044928
- daf341df35

Thanks @mistakenot!
